File: 2024/python/day1/part1.py
from pathlib import Path
import re
from typing import List
import argparse


def main(input_file: Path):
    with input_file.open() as f:
        input_data: List[str] = f.readlines()
        result = part1(input_data)
    print(result)


def part1(input: List[str]) -> int:
    left_numbers: List[int] = []
    right_numbers: List[int] = []
    input_pattern = r"^(\d+)\s+(\d+)$"
    for l in input:
        numbers = re.findall(input_pattern, l)
        if not numbers:
            raise RuntimeError("Could not obtain numbers from input")
        left_number, right_number = numbers[0]
        left_numbers.append(int(left_number))
        right_numbers.append(int(right_number))
    left_numbers.sort()
    right_numbers.sort()
    distance = 0
    for left, right in zip(left_numbers, right_numbers):
        distance += get_distance(left, right)
    return distance


def get_distance(left_number: int, right_number: int):
    return abs(left_number - right_number)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("input_file", type=Path)
    args = parser.parse_args()
    main(input_file=args.input_file)

File: 2024/rust/day1/src/bin/part1.rs
use std::env;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;
use std::iter::zip;

fn main() {
    let args: Vec<String> = env::args().collect();
    let input_path = &args[1];
    let input_file = File::open(input_path).expect("Can't open file!");
    let results = part1(input_file);
    println!("The total distance is: {}", results);
}

fn part1(input_file: File) -> i32 {
    let reader = BufReader::new(input_file);
    let mut left_numbers: Vec<i32> = vec![];
    let mut right_numbers: Vec<i32> = vec![];
    for line in reader.lines() {
        let input = line.expect("Could not read lines of input file");
        let mut numbers = input.split_whitespace();
        let left_number = numbers
            .next()
            .expect("Could not extract left number")
            .parse()
            .expect("Could not parse left number");
        let right_number = numbers
            .next()
            .expect("Could not extract right number")
            .parse()
            .expect("Could not parse right number");
        left_numbers.push(left_number);
        right_numbers.push(right_number);
    }
    left_numbers.sort();
    right_numbers.sort();
    let mut distance = 0;
    for (left, right) in zip(left_numbers, right_numbers) {
        distance += (left - right).abs()
    }
    distance
}
